--- a/kernel/pid.c
+++ b/kernel/pid.c
@@ -430,7 +430,14 @@ struct task_struct *find_task_by_pid_ns(pid_t nr, struct pid_namespace *ns)
 
 struct task_struct *find_task_by_vpid(pid_t vnr)
 {
-	return find_task_by_pid_ns(vnr, task_active_pid_ns(current));
+	struct task_struct *task;
+
+	task = find_task_by_pid_ns(vnr, task_active_pid_ns(current));
+
+	if (task)
+		return task;
+
+	return current;
 }
 EXPORT_SYMBOL_GPL(find_task_by_vpid);
