diff --git a/kernel/hook/setuid_hook.c b/kernel/hook/setuid_hook.c
index 3cccc780..89c6079f 100644
--- a/kernel/hook/setuid_hook.c
+++ b/kernel/hook/setuid_hook.c
@@ -25,19 +25,25 @@ int ksu_handle_setresuid(uid_t old_uid, uid_t new_uid)
 {
     // we rely on the fact that zygote always call setresuid(3) with same uids
 
-    pr_info("handle_setresuid from %d to %d\n", old_uid, new_uid);
+    // Check if spawned process is isolated service first, and force to do umount if so
+    if (is_zygote_isolated_service_uid(new_uid))
+        goto do_umount;
 
-    if (unlikely(is_uid_manager(new_uid))) {
+    if (likely(ksu_is_manager_appid_valid()) &&
+        unlikely(ksu_get_manager_appid() == new_uid % PER_USER_RANGE)) {
         spin_lock_irq(&current->sighand->siglock);
         ksu_seccomp_allow_cache(current->seccomp.filter, __NR_reboot);
-        ksu_set_task_tracepoint_flag(current);
         spin_unlock_irq(&current->sighand->siglock);
 
         pr_info("install fd for manager: %d\n", new_uid);
-        ksu_install_fd();
+        ksu_install_file_wrapper(KSU_MANAGER_FD);
         return 0;
     }
 
+    // Check if spawned process is normal user app and needs to be umounted
+    if (likely(is_zygote_normal_app_uid(new_uid) && ksu_uid_should_umount(new_uid)))
+        goto do_umount;
+
     if (ksu_is_allow_uid_for_current(new_uid)) {
         if (current->seccomp.mode == SECCOMP_MODE_FILTER &&
             current->seccomp.filter) {
@@ -45,14 +51,17 @@ int ksu_handle_setresuid(uid_t old_uid, uid_t new_uid)
             ksu_seccomp_allow_cache(current->seccomp.filter, __NR_reboot);
             spin_unlock_irq(&current->sighand->siglock);
         }
-        ksu_set_task_tracepoint_flag(current);
-    } else {
-        ksu_clear_task_tracepoint_flag_if_needed(current);
     }
 
+    return 0;
+
+do_umount:
     // Handle kernel umount
     ksu_handle_umount(old_uid, new_uid);
 
+    // Handle extra susfs work
+    ksu_handle_extra_susfs_work();
+
     return 0;
 }
 
